Specialize map and map! for BitArrays

mbauman opened the pull request:
Alright, next step in getting IntSets to use BitArrays: a faster map.

Specializing map is even more important for bitarrays than it is for generic arrays since there can be a 64x+ speedup by working at the level of Int64 instead of looping bit-by-bit. There's no other way to perform operations at the chunk-level without either exporting more functions or mucking around with BitArray's internal fields (See mbauman/IntSets.jl#1).

This adds functor types (a la reduce.jl) that allow mapping across BitArrays 64-bits at a time. Performance of `map(&, x, y)` is now ~300x faster and very competitive with `x & y` for large arrays. In-place `map!` is even faster. It's a little strange to do bitwise/logical operations using comparisons, but it is consistent and works… and otherwise, a whole slew of new names would need to be exported (e.g., nor, nand, implies, etc.).  Also cool: this gives a very performant in-place inverse method — `map!(~, p)` — without any temporaries or new function exports.

Demo:

``` julia
### BEFORE ###
julia> p = bitrand(10^8); q = bitrand(10^8); r = bitrand(10^8);

julia> map(&, p, q); @time map(&, p, q);
elapsed time: 4.875528337 seconds (12513944 bytes allocated)

julia> map!(&, r, p, q); @time map!(&, r, p, q);
elapsed time: 9.50449699 seconds (1599991904 bytes allocated, 16.76% gc time)

julia> p & q; @time p & q;
elapsed time: 0.013313429 seconds (12500176 bytes allocated)

julia> map!(~, p); @time map!(~, p);
elapsed time: 9.311276771 seconds (1599991904 bytes allocated, 17.19% gc time)

### AFTER ###
julia> map(&, p, q); @time map(&, p, q);
elapsed time: 0.016313608 seconds (12513904 bytes allocated)

julia> map!(&, r, p, q); @time map!(&, r, p, q);
elapsed time: 0.011889581 seconds (80 bytes allocated)

julia> p & q; @time p & q;
elapsed time: 0.01494325 seconds (12500176 bytes allocated)

julia> map!(~, p); @time map!(~, p);
elapsed time: 0.008248346 seconds (80 bytes allocated)

julia> f!(r,p,q) = for i=1:length(p); r[i] = p[i] & q[i]; end
       f!(r,p,q); @time f!(r, p, q); # Much faster than manual devectorization, too
elapsed time: 1.299135844 seconds (80 bytes allocated)
```

There are a few other changes here:
- The old map definitions in bitarray.jl weren't really bitarray defitions.  They were defining methods for `Union(StridedArray, BitArray)`… but `BitArray <: StridedArray`.  It seems like they belong much better in abstractarray.jl, and can be defined for all abstractarrays.
- There were also definitions for `map!` with combinations of numbers.  I think those should be deprecated in favor of `broadcast!`.

cc @carlobaldassi again - thanks!

mbauman committed:
Deprecate map!(…, ::Number, …) for broadcast!

Also rename `Union(StridedArray, BitArray)` to just StridedArray since BitArray <: StridedArray.

mbauman committed:
Move map!(f, ::StridedArray…) to abstractarray.jl

Since these methods should work for all writable abstract arrays, broaden the method definition to map!(f, ::AbstractArray…).

mbauman committed:
Specialize map and map! for BitArrays

Specializing map is even more important for bitarrays than it is for generic
arrays since there can be a 64x speedup by working at the level of Int64
instead of looping bit-by-bit. There's no other way to perform operations at the chunk-level without either exporting more functions or mucking around with BitArray's internal fields.

This adds functor types that allow mapping across BitArrays 64-bits at a time. Performance of `map(&, x, y)` is now 300x faster and very competitive with `x & y`. In-place `map!` is even faster. The one downside here is that bitwise operations are not typically performed using comparisons... but otherwise a whole slew of new names would need to be exported (e.g., nor, nand, implies, etc.).

carlobaldassi commented on base/bitarray.jl:
```diff
@@ -1473,49 +1473,106 @@ maximum(B::BitArray) = isempty(B) ? error("argument must be non-empty") : any(B)
+##  p = TTFF                          ##  p = TTFF                          ##
+##  q = TFTF    function  name        ##  q = TFTF    function  name        ##
+##  --------------------------------  ##  --------------------------------- ##
+##      TTTT    (true)    -           ##      FFFF    (false)   -           ##
```
these two can be associated with `one` and `zero` functions

mbauman replied:
I'm not sure it's worth taking the time to check for `one` and `zero` in specializing the functor types — there's already `fill!` that does this sort of operation very well.

carlobaldassi commented on base/bitarray.jl:
```diff
@@ -1473,49 +1473,106 @@ maximum(B::BitArray) = isempty(B) ? error("argument must be non-empty") : any(B)
+    t != nothing ? map!(t, dest, A, B) : invoke(map!, (Callable, AbstractArray, AbstractArray, AbstractArray), f, dest, A, B)
+end
+
+function map!(f, dest::BitArray, A::BitArray)
```
I think the first argument here should have a `Func{1}` type signature, no?

carlobaldassi replied:
Ah no sorry I see it wouldn't work like that.

mbauman replied:
No, your first hunch here was right.  I initially had it more strongly typed like that but then removed it to be more like reduce.jl.

carlobaldassi commented on base/bitarray.jl:
```diff
@@ -1473,49 +1473,106 @@ maximum(B::BitArray) = isempty(B) ? error("argument must be non-empty") : any(B)
+    dest.chunks[end] = f(A.chunks[end]) & _msk_end(A)
+    dest
+end
+function map!(f, dest::BitArray, A::BitArray, B::BitArray)
```
also `f::Func{2}` here?

carlobaldassi commented on base/bitarray.jl:
```diff
@@ -1473,49 +1473,106 @@ maximum(B::BitArray) = isempty(B) ? error("argument must be non-empty") : any(B)
+immutable NotFun <: Func{1} end
+call(::NotFun, x) = ~x
+
+# Use IdFun, AndFun, and OrFun from reduce.jl
```
We might as well have all the definitions in one place, and just add a note in reduce.jl that says to look in bitarrays.jl

carlobaldassi replied:
Or rather, put all definitions into a functors.jl file?

mbauman replied:
Yes, a functors file is a good idea.  You know, with that change and the addition of more `specialize_*` functions, it'd be trivial to opt-in to using functors.  Especially if there was an `immutable UnspecializedFunc{N,T} <: Func{N}; f::T; end` which would just fall-back to the stored function.  It'd dramatically simplify using Functors - there wouldn't need to be any branches written in the reduce.jl code.  I'll take a look into doing that tonight.

carlobaldassi commented on base/bitarray.jl:
```diff
@@ -1473,49 +1473,106 @@ maximum(B::BitArray) = isempty(B) ? error("argument must be non-empty") : any(B)
+# Use IdFun, AndFun, and OrFun from reduce.jl
+
+immutable POrNotQFun <: Func{2} end
+call(::POrNotQFun, x, y) = x | ~y
```
If we defined `call(::Type{POrNotFun}, x, y)` rather than `call(::POrNotQFun, x, y)`, and returned `POrNotFun` rather than `POrNotFun()` from `specialized_binary`, and of course analogous definitions for the other functions, then we could use `Func{1}`/`Func{2}` in the type signature of `map`, which I think would be safer/clearer.

carlobaldassi commented:
This is great! I also much needed something like this in my own code, where I had to use BitArrays internal fields in order to avoid allocation.

It's a bit unfortunate that we can't expose all possible 2-argument functions for the lack of a corresponding name in Base.

carlobaldassi commented:
I don't think we should deprecate the `map` versions which use `Number` arguments. I think that they follow from making scalars iterable. Maybe they should just be added in abstractarray.jl.

mbauman commented:
I decided to deprecate `map(… ::Number …)` for the following reasons:

* As written, it only works with one- or two-argument functions.  `map!(+, zeros(2), 1, [1,2])` works but `map!(+, zeros(2), 1, [1,2], [3,4])` doesn't.

* The broadcast! function already does this much more robustly.

* Documentation for map doesn't say anything about broadcasting scalars.

But if this is at all controversial I'll happily split it out into a separate issue.

mbauman committed:
Refactor Functors into their own file

Add a new BitFunc abstract type to group Functors that behave in the same bit-wise manner when applied to individual bits as well as integers, allowing them to be used on the BitArray chunks.

Add methods for Functor specialization-by-function value. When no Functor matches, an UnspecializedFunc is returned that simply wraps the function. This should make dispatch simpler when using functors:

```
foo(f::Callable, ...) = foo(specialized_(f), ...)
function foo(f, ...)
    # actual implementation here
end
```

mbauman commented:
I've refactored Functors as suggested.  See what you think.  Also now cc @lindahua since this is getting into the reduce codebase.

carlobaldassi commented:
Travis error is perplexing but seems unrelated.
The refactoring seems good to me, let's also wait for @lindahua's opinion.

about this:
> I'm not sure it's worth taking the time to check for `one` and `zero` in specializing the functor types — there's already `fill!` that does this sort of operation very well.

it's nice for completeness, and after all even mapping the identity does not seem all that useful. I don't think the time spent in the check would be in any way significant.

Anyway, the way I see it, one of the nice things of this PR besides the non-allocating maps is that it also allows to work-around the limitations of `map` when applied to BitArrays. The reason `map` can't directly use the trick of `findnext`, i.e. evaluating the function only a bunch of times and reusing the results, is of course that functions may have state (e.g. `rand`). But if the user knows that her function does not have state, than she can map it to one of the 16 possible functions which transform 2 bits into 1 bit. In other words, one could define a `bitmap` function (disregard the silly name) which evaluates `f(true, true)`, `f(true, false)` etc. and then uses the table which you wrote in the comments. Except some functions are missing now, which is why I'd actually like to see them all implemented somehow. Anyway this doesn't need to be completed immediately, I think we can merge the PR and do that at a later time.

mbauman commented:
The travis error looked like it stemmed from a Sets worker failing.  I've rebased onto the latest master to see if we can get around these intermittent failures.

Unfortunately, `one` and `zero` are unary operations, so they won't work to fill out the binary operators.  I can add them as `BitFunc{1}`s, though, and check for them in the unary case.

Here's a thought: What about a generic `BitFunctor` type:

```julia
immutable BitFunctor{TT,TF,FT,FF} <: Func{2} end
BitFunctor(f::Function) = is(f, &) ? BitFunctor{true, false, false false}() : …
BitFunctor(; TT=error("all truth values must be specified"), TF=error(), FT=error(), FF=error()) = BitFunctor{TT,TF,FT,FF}()
call(::BitFunctor{true, false, false, false}, x, y) = x & y
…
```

This way, even if the function doesn't exist in Base, you can still map the operation performantly.

carlobaldassi commented:
> Unfortunately, `one` and `zero` are unary operations, so they won't work to fill out the binary operators.

I was simply thinking about this

```
immutable BinaryOneFun <: BitFunc{2} end
call(::BinaryOneFun, x, y) = _msk64

immutable BinaryZeroFun <: BitFunc{2} end
call(::ImpliesFun, x, y) = uint64(0)
```

etc. (same for unary)

> Here's a thought: What about a generic `BitFunctor` type:

I like that. Maybe avoid keyword arguments for performance: we'd just need `BitFunctor(::Function)` and

```
BitFunctor(TT,TF,FT,FF) = BitFunctor{Bool(TT),Bool(TF),Bool(FT),Bool(FF)}()`
```

mbauman committed:
Rework bitwise functors into just 2 parametric types

This has several advantages: it reduces the number of names required to get good performance, it makes the truth tables much more obvious, and it allows specialization for all 16 possible bitwise binary operators (not just the ones that happen to be possible with existing Base functions).

carlobaldassi commented:
@mbauman I see you have implemented the last changes - which look great. Is there anything else to do here? Tests pass and the merge conflict is trivial.

For the record, I tried the branch since I wanted to check type inference, which does not seem worse then before this PR (e.g. `map(&, ::BitArray, ::BitArray)` return type is inferred as `AbstractArray{T,N}`, and I doubt it could be better than that). BTW this is now probably the most important advantage of the generic functions over the `map` versions.

mbauman commented:
Yup! This is good to go as far as I'm concerned.  Thanks again for all the feedback!

simonster commented on base/functors.jl:
```diff
@@ -0,0 +1,135 @@
+immutable UnspecializedFun{N,T<:Callable} <: Func{N}
+    f::T
+end
+call{N,T}(::Type{UnspecializedFun{N}}, f::T) = UnspecializedFun{N,T}(f)
```
This causes performance problems with `map(Int, BitArray)` because `specialized_bitwise_unary` returns an `UnspecializedFun{N,DataType}` instead of an `UnspecializedFun{N,Type{Int}}`. There's an obvious fix to make it specialize properly for DataTypes, but I wasn't sure why this type was necessary. Why can't `specialized_unary` etc. just return `f` if there's no corresponding specialized functor?
